### PR TITLE
Archive parity c driver and tests, #632 and #621

### PR DESCRIPTION
**Pull request by mikeb01**
This PR implements multi-destination subscriptions and session specific subscriptions in order to support the archive with the C media driver.  This includes a number of other fixes and enhancements to provide the supported functionality.

- int64 counter map, similar to the int to int counter map from Agrona.
- int64 to tagged ptr map, which allows additional information to be stored alongside the ptr.  Useful for reducing the number of data-dependent loads.
- Start extracting some common map functions to an aeron_map.{h,c}.
- Add a `remove_if` function to the ptr maps that allows elements to be removed from the map based on a predicate.
- Add unit tests for data packet dispatcher and publication image.
- Allow for more rapid open/close of subscription on the same endpoint port.  Have the client handle a warning error, meaning the operation can be re-tried if there is a clash between ports (window is much narrower than before).
- Get sanitized build passing.
- Ensure ip/port pairs are formatted in URI-style for IPv6 addresses (e.g. `[::1]:1234`).
- Add missing support for tags which using session specific publications.
- Get all MDS, Session Specific Subscription and Archive system tests passing against the C media driver.

### CODE REVIEW

File: aeron-driver/src/main/c/aeron_driver_conductor.c [hunk 1]
```diff
@@ -419,6 +487,33 @@ int aeron_confirm_publication_match(
...
+    if (link->has_session_id)
+    {
+        aeron_receive_channel_endpoint_decref_to_stream_and_session(
+            endpoint, link->stream_id, link->session_id);
+    }
+    else
+    {
+        aeron_receive_channel_endpoint_decref_to_stream(endpoint, link->stream_id);
+    }
+//    if (AERON_RECEIVE_CHANNEL_ENDPOINT_STATUS_CLOSING == endpoint->conductor_fields.status)
```
Comment (mjpt777) [inline aeron-driver/src/main/c/aeron_driver_conductor.c:504]: Why not remove commented code?

File: aeron-system-tests/src/test/java/io/aeron/ImageAvailabilityTest.java [hunk 2]
```diff
@@ -37,18 +39,21 @@
     private static List<String> channels()
     {
         return asList(
-            "aeron:ipc?term-length=64k",
-            "aeron:udp?endpoint=localhost:24325|term-length=64k",
+//            "aeron:ipc?term-length=64k",
+//            "aeron:udp?endpoint=localhost:24325|term-length=64k",
```
Comment (mjpt777) [inline aeron-system-tests/src/test/java/io/aeron/ImageAvailabilityTest.java:43]: Should these be added back?

File: aeron-driver/src/test/java/io/aeron/driver/DriverConductorTest.java [hunk 3]
```diff
@@ -348,6 +348,20 @@ public void shouldBeAbleToRemoveMultipleStreams()
...
+    @Test
+    public void shouldBeAbleToAddWithStreamIdTag()
+    {
+        final long id1 = driverProxy.addPublication(CHANNEL_4001 + "|tags=1,2", STREAM_ID_1);
+        final long id2 = driverProxy.addPublication(CHANNEL_4002 + "|session-id=tag:2", STREAM_ID_1);
+
+        doWorkUntil(
+            () -> (CLIENT_LIVENESS_TIMEOUT_NS * 2 + PUBLICATION_LINGER_TIMEOUT_NS * 2) - nanoClock.nanoTime() <= 0);
+
+//        verify(senderProxy, times(4)).removeNetworkPublication(any());
```
Comment (mjpt777) [inline aeron-driver/src/test/java/io/aeron/driver/DriverConductorTest.java:361]: Is this still needed?

File: aeron-driver/src/main/c/aeron_driver_conductor.c (hunk 1 above)
Comment (mikeb01) [reply aeron-driver/src/main/c/aeron_driver_conductor.c:504]: Removed

File: aeron-system-tests/src/test/java/io/aeron/ImageAvailabilityTest.java (hunk 2 above)
Comment (mikeb01) [reply aeron-system-tests/src/test/java/io/aeron/ImageAvailabilityTest.java:43]: Yes, done.

File: aeron-driver/src/test/java/io/aeron/driver/DriverConductorTest.java (hunk 3 above)
Comment (mikeb01) [reply aeron-driver/src/test/java/io/aeron/driver/DriverConductorTest.java:361]: No, test removed.  Was used to explore behaviour on the Java side.

**Pull request merged**